[FIXED JENKINS-31055] Make Node implement Saveable

stephenc opened the pull request:
- Also tidy up some of the mess logic in both #1866 and #1860 

@reviewbybees

stephenc committed:
[FIXED JENKINS-31055] Make Node implement Saveable

- Also tidy up some of the mess logic in both #1866 and #1860

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

amuniz commented:
:bee:

oleg-nenashev commented on changelog.html:
```diff
@@ -56,6 +56,9 @@
 <div id="trunk" style="display:none"><!--=TRUNK-BEGIN=-->
 <ul class=image>
   <li class=>
```
Delete this garbage stub

stephenc replied:
Nope that stub it a placeholder that @kohsuke adds every release

recena replied:
A very fragile placeholder :dizzy_face:

stephenc replied:
removing it makes merging harder if there are competing pull requests that land first

jtnord commented on core/src/main/java/jenkins/model/Nodes.java:
```diff
@@ -166,17 +168,27 @@ private void persistNode(final @Nonnull Node node)  throws IOException {
+            // can never happen
+            exists = false;
+        }
+        if (exists) {
```
this is no longer thread safe - as the node may have been removed.
an update to a removed node should fail. (return `false`)

stephenc replied:
It is more thread safe than before. in 1.634 the node can be removed before getting the queue lock.

We cannot hold a lock while performing IO as that will drown the queue.

This makes the update method as thread safe as `addNode` has been

rsandell commented:
:bee:

jtnord commented on core/src/main/java/hudson/model/Node.java:
```diff
@@ -115,6 +118,18 @@ public boolean isHoldOffLaunchUntilSave() {
+        final Jenkins jenkins = Jenkins.getInstance();
+        if (jenkins != null) {
+            jenkins.updateNode(this);
+        }
```
surely this should throw an exception if Jenkins is null?
otherwise the caller will think that the `save` operation has succeeded.

stephenc replied:
The default implementation of `Saveable.save()` for most cases will be a no-op if there is no place to save the configuration. Without a `Jenkins` there is no JENKINS_HOME and thus there is no config file to save, so the "normal" thing is to just do a no-op

jglick replied:
It would mean something was scheduled into a background thread that had the bad luck to run while Jenkins was shutting down, so the edit gets lost, oh well.

jtnord commented:
:bug: saves are silently ignored if jenkins is shutting down - and lack of thread safety (resurecting dead nodes)

stephenc commented:
@jtnord during shut-down, at the point where `Jenkins.getInstane() == null` there could be a lot of changes to Nodes resulting from the disconnects etc that should not be persisted to disk as they risk corrupting the state during startup.

jtnord commented:
:bee: based on responses

> during shut-down, at the point where `Jenkins.getInstane() == null` there could be a lot of changes to Nodes resulting from the disconnects etc that should not be persisted to disk as they risk corrupting the state during startup.

I feel like that should be taken care of in the callers of this function rather than here.  Throwing an `IOException` (`IllegalStateException`) would seem better as there may be code that does want to save something and this could be reported incorrectly causing some as yet undiscovered bug.

stephenc commented:
@jtnord so the instance is only set to null after the cleanUp lifecycle has completed, the timers have been shut down, the tcp agent shut down, the plugin manager shut down, the queue saved, and up to 10 seconds per running job threads to complete and the loggers all released... is there really all that much that can be left where there might be something needing to save at that point?

stephenc commented:
@reviewbybees done

reviewbybees commented:
This pull request has completed our internal processes and we now respectfully request the maintainers of this repository to consider our proposal contained within this pull request for merging.

KostyaSha commented:
Test cases? `System.exit(0)` from extension points/listeners should do the trick with force shutdown.

jglick commented on core/src/main/java/hudson/model/Node.java:
```diff
@@ -115,6 +118,18 @@ public boolean isHoldOffLaunchUntilSave() {
 
     /**
+     * {@inheritDoc}
+     * @since 1.635
```
Careful—can be easily forgotten if merging does not happen as quickly as you had hoped. Prefer to use `TODO` or `FIXME`. @kohsuke [uses `1.THU`](https://github.com/jenkinsci/jenkins/blob/11c81cacf5776415cd68ee26f32e5d28549a0322/core/src/main/java/jenkins/security/s2m/AdminCallableMonitor.java#L19) and do not ask me [what it means](http://www.acronymfinder.com/THU.html). :-)

jglick commented on core/src/main/java/jenkins/model/Nodes.java:
```diff
@@ -166,17 +168,27 @@ private void persistNode(final @Nonnull Node node)  throws IOException {
                 }
             });
+        } catch (Exception e) {
+            // can never happen
```
Well, it could be a `RuntimeException`.

jglick commented:
The test failure is random. (This CI job is not yet set up to rerun flaky tests.)

stephenc committed:
[JENKINS-31055] Finessing the exception handling

stephenc committed:
[JENKINS-31055] Replacing @since tag with placeholder to be restored on merge

stephenc commented:
@KostyaSha a `System.exit(0)` does not invoke the default termination lifecycle (which is why killing Jenkins or a Jenkins crash does not persist the Queue whereas an orderly shutdown using the Servlet Container lifecycle does)

The only way you can get to the `Node.save()` being a no-op is through an orderly shutdown where a background thread has taken more than 10 seconds to stop (as all the stuff doing background work on `Node` should be in `Computer.threadpoolForRemoting` which is the thread pool that the orderly shutdown lifecycle waits for)...

Only after all of that has completed is `Jenkins.getInstance() == null`

I fail to see that it could even be possible to save anything at that point as so much of the Jenkins code base assumes that `Jenkins.getInstance() != null` so in all likelihood, even if we cached the config file location, the actual `XmlFile.write` would fail anyway....

Currently no plugin can be calling `Node.save()` so they would be calling `Jenkins.getInstance().save()` or `Jenkins.getActiveInstance().save()` to achieve the same end result...

We need to have `Node.save()` be a no-op when the node is not in the list of Jenkins nodes so that jenkins is in control of defining exactly what nodes it has... thus the persist only occurs if the node instance is in the node list or if there is a node list at all

stephenc committed:
[JENKINS-31055] The intent of Node.save() is a no-op when not in Jenkins list of nodes

stephenc committed:
[JENKINS-31055] Note theoretical race condtion

Seems rather unlikely, but none the less it is a possibility. Not sure how we should handle this though:

* We do not want to hold the Queue lock while performing I/O as that will grind everything to a halt
* We do not want a big fat lock on all nodes
* If we have a per-node lock then the save everything method may have to grab them all

In short, it gets ugly real fast.

For now, I believe that the risk is very low as releasing a lock does not intrinsically deschedule a currently running thread so the likely course of events is that the thread will release the lock and at least start the save before a competing thread could acquire the lock update remove the node release the lock and delete the node's directory

stephenc commented:
I believe I have addressed all comments... my intent is to merge this by tomorrow unless I hear otherwise

KostyaSha commented:
So comment is not write tests?

stephenc commented:
Existing tests cover the reasonable code paths. The only way to hit an unreasonable code path is in very strange shutdown scenarios where the test would be subject to timing issues (or flakey) and where the current behaviour can be perfectly reasoned about by code inspection.

So yeah, no new tests
